### PR TITLE
feat: Add Chat Message History for Firestore.

### PR DESCRIPTION
**Pull request by JU-2094**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [X] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/langchain-google-firestore-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Add Chat Message History for Firestore. This PR was pulled from Document Loader PR - https://github.com/googleapis/langchain-google-firestore-python/pull/9

### CODE REVIEW

**Comment by JU-2094** [comment]
I'm planning to first merge the Document Loader PR -- https://github.com/googleapis/langchain-google-firestore-python/pull/9

And then merge this PR that adds the Chat Message History.

**Review by kurtisvg — CHANGES_REQUESTED** [review]
This PR is pretty huge -- maybe it would be better to separate some of the utilities into their own PRs to shrink a bit?

**Comment by JU-2094** [comment]
> This PR is pretty huge -- maybe it would be better to separate some of the utilities into their own PRs to shrink a bit?

@kurtisvg This is because this branch was based from https://github.com/googleapis/langchain-google-firestore-python/pull/9 and it's sharing files. But in reality this PR it's less files (the chat history only). The utility is used only at the document loader.

I had in mind to merge first https://github.com/googleapis/langchain-google-firestore-python/pull/9 then sync , which should remove files here. Sorry I had in mind some 'fig' approach here. But I don't know how to indicate there's a base PR. I did it for the shared files e.g. __init__ , etc.

The files were `chat_message_history.py` and `test_chat_message_history.py` .
The converter is not used in the `chat_message_history.py`

As an alternative I could remove the files from this PR.  - Please let me know if you prefer that. (I would be removing all Document Loader from  https://github.com/googleapis/langchain-google-firestore-python/pull/9) except shared files.

File: src/langchain_google_firestore/chat_message_history.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        client: Optional[Client] = None,
+    ) -> None:
+        """Chat Message History for Google Cloud Firestore.
+        Args:
+            session_id: Arbitrary key that is used to store the messages of a single
+                chat session. This is the document_path of a document.
+            collection: The single `/`-delimited path to a Firestore collection.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        try:
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py:57]: Done.  I also did it for the Doc Loader PR.

File: src/langchain_google_firestore/chat_message_history.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        client_info = DEFAULT_CLIENT_INFO
+        client_info.user_agent = USER_AGENT
+        if client:
+            self.client = client
+            self.client._user_agent = USER_AGENT
+        else:
+            self.client = firestore.Client(client_info=client_info)
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py:71]: DONE

File: src/langchain_google_firestore/chat_message_history.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+from langchain_core.messages import (
+    BaseMessage,
+    messages_from_dict,
+)
+
+USER_AGENT = "LangChain"
+DEFAULT_COLLECTION = "ChatHistory"
+IMPORT_ERROR_MSG = (
+    "`firestore` package not found, please run `pip3 install google-cloud-firestore`"
+)
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py:37]: DONE

**Comment by kurtisvg** [comment]
> > This PR is pretty huge -- maybe it would be better to separate some of the utilities into their own PRs to shrink a bit?
> 
> @kurtisvg This is because this branch was based from #9 and it's sharing files. But in reality this PR it's less files (the chat history only). The utility is used only at the document loader.
> 
> I had in mind to merge first #9 then sync , which should remove files here. Sorry I had in mind some 'fig' approach here. But I don't know how to indicate there's a base PR. I did it for the shared files e.g. **init** , etc.
> 
> The files were `chat_message_history.py` and `test_chat_message_history.py` . The converter is not used in the `chat_message_history.py`
> 
> As an alternative I could remove the files from this PR. - Please let me know if you prefer that. (I would be removing all Document Loader from #9) except shared files.

Ok - I misread that. You can tell GitHub which branch is the base PR and it'll adjust accordingly. I just updated this branch to point to document loaders as the base. 

I think in general it's preferable to have the PRs to have as minimal overlap as possible so they can be reviewed/modified concurrently. If there are shared files, I generally try to pull those out into it's own PR for the same reason.

File: src/langchain_google_firestore/utility/document_converter.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        else:
+            set_page_fields = set()
+        if metadata_fields:
+            set_metadata_fields = set(metadata_fields)
+        else:
+            set_metadata_fields = set()
+        shared_keys = set_metadata_fields & set_page_fields
+
+        page_content = {}
+        for k in sorted(shared_keys):
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py]: The code don't need this, but for the testing is failing because the dictionary keys are not sorted and constructing the `page_content` results in different strings. Which causes tests to fail intermittently.

e.g.
"{'f1':1, 'f2':2}"   -- sometimes is "{'f2':2, 'f1':1}"

I could remove the sort but then tests would be flaky.

File: src/langchain_google_firestore/utility/document_converter.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    @staticmethod
+    def _convertFromLangChain(val: Any, client: Client) -> Any:
+        try:
+            from google.cloud.firestore_v1.document import DocumentReference
+            from google.cloud.firestore_v1._helpers import GeoPoint
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        val_converted = val
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py:160]: I'm skipping the else statement with this.  Otherwise I just can add an else at the end and do this same assignment.

File: src/langchain_google_firestore/utility/document_converter.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                for k in keys
+            }
+        else:
+            page_content.update(
+                {
+                    k: DocumentConverter._convertFromFirestore(data_doc.pop(k))
+                    for k in sorted(set_page_fields - shared_keys)
+                    if k in data_doc
+                }
+            )
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py]: ### About the `pop`
* With the `pop` I don't need to keep track of which keys I already assigned to `page_content` or `metadata`. 
* I can iterate once through the `metadata_fields` and remove those values so I don't keep track of the assigned keys and then add it to the `page_content`. If I had to keep track of the keys more conditionals are needed.
* When using `doc.to_dict()` we get a copy and not a reference so we are not modifying the original document.
* If `metadata_fields` is None/empty I need to set the remaining keys-values that where not assigned to `page_content`; removing the keys-values from the data is an alternative from tracking which keys I assigned and adding a conditional.
** A similar logic applies if `page_content_fields` is empty. I need to write the `metadata_fields`, keep track of the items and assign the remaining to `page_content`.

### For the sets (shared keys, and rest)
* If `page_content_fields` and `metadata_fields` share keys.  Both `page_content` and `metadata` will have the information.  
* The rest is to assign the remaining key-values. I don't like requiring the `sorted` for the testing and I'm reducing the number of elements before the sort. Otherwise we could iterate, with the `pop` we are sure we won't be adding duplicates.

* The behaviors when the metadata_fields or page_content_fields , and the testing flakiness led to this approach. Still there are similar iterations. I can iterate once for metadata_fields, pop the key-values and then go for `page_content`. But again if there are page_content_fields and metadata_fields is None. The remaining values unassigned go to `metadata`

File: src/langchain_google_firestore/chat_message_history.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            self.client = client
+            self.client._user_agent = USER_AGENT
+        else:
+            self.client = firestore.Client(client_info=client_info)
+        self.session_id = session_id
+        self.doc_ref = self.client.collection(collection).document(session_id)
+        self.messages: List[BaseMessage] = []
+        self.load_messages()
+
+    def load_messages(self) -> None:
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py:77]: SGTM, changed.

File: src/langchain_google_firestore/chat_message_history.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    def upsert_messages(self) -> None:
+        self.doc_ref.set({"messages": self._encode_messages(self.messages)})
+
+    def clear(self) -> None:
+        self.messages = []
+        self.doc_ref.delete()
+
+    def _encode_messages(self, messages: List[BaseMessage]) -> List[str]:
+        return [str.encode(m.json()) for m in messages]
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py:96]: SG, moved to another class.

File: src/langchain_google_firestore/document_loader.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+    def lazy_load(self) -> Iterator[Document]:
+        """A lazy loader for Documents."""
+        try:
+            from google.cloud.firestore_v1.document import DocumentReference
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        if isinstance(self.source, DocumentReference):
+            self.source._client._client_info.user_agent = USER_AGENT
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:93]: Yes, it's a different client and to keep tracking of the feature usage I need to set this Header.

This is overriding the USER_AGENT header, which is informational.

File: src/langchain_google_firestore/document_loader.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        db_batch = self.client.batch()
+
+        if document_ids and (len(document_ids) != len(documents)):
+            raise ValueError("Document ids and docs must have the same size")
+
+        if document_ids:
+            docs_list = tuple(zip(documents, document_ids))
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:174]: I use the len for creating the batches in the method _batched. And for zip objects, there is no way to count the length. zip has no len() . We can re-cast it to a list or tuple, which does have a len()

File: src/langchain_google_firestore/document_loader.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        else:
+            docs_list = documents
+
+        for batch in self._batched(docs_list, WRITE_BATCH_SIZE):
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:178]: I don't think so. This method returns a slice of the data
I did some local testing and is not behaving as I expect
```
    # islice('ABCDEFG', 2) --> A B
    # islice('ABCDEFG', 2, 4) --> C D
    # islice('ABCDEFG', 2, None) --> C D E F G
    # islice('ABCDEFG', 0, None, 2) --> A C E G
```

```
>>> docs = zip(['a','b','c','d'],[1,2,3,4])
>>> for batch in itertools.islice(docs,2):
...   for e in batch:
...     print(e)
...   print(batch)
...   print('--')
...
a
1
('a', 1)
--
b
2
('b', 2)
--
```

While I want batches of size N . Not a single slice. I searched for this, there are implementation using islice as part of another algorithm that creates batches. Which looks more complicated. 

There is another package called more_itertools that does what I need https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.chunked
However I was not certain of the packages installation process. -- If you see this more fitting we can install this package and use that method.

File: src/langchain_google_firestore/document_loader.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        """
+        try:
+            from google.cloud.firestore_v1.document import DocumentReference
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        db_batch = self.client.batch()
+
+        if document_ids and (len(document_ids) != len(documents)):
+            raise ValueError("Document ids and docs must have the same size")
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:171]: DONE

File: src/langchain_google_firestore/document_loader.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+
+        if document_ids:
+            docs_list = tuple(zip(documents, document_ids))
+        else:
+            docs_list = documents
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:176]: I get in the same situation with the `zip` that there is no len() for creating the batches.
I can complicate the _batched logic with itertool.islice to do batching  or install the more_itertools package to use chunked.

I'm OK with either approach or keep current.

File: src/langchain_google_firestore/document_loader.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:178]: I will opt to install this package to address other comments

File: src/langchain_google_firestore/document_loader.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:176]: Opted to use more_itertools and itertools.

File: src/langchain_google_firestore/document_loader.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+
+        for batch in self._batched(docs_list, WRITE_BATCH_SIZE):
+            for elem in batch:
+                if document_ids:
+                    doc = elem[0]
+                    doc_id = elem[1]
+                else:
+                    doc = elem
+                    doc_id = None
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:185]: DONE

File: src/langchain_google_firestore/document_loader.py [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+                if self.collection:
+                    doc_ref = self.client.collection(self.collection).document()
+                elif doc_id:
+                    doc_ref = DocumentReference(*doc_id.split("/"), client=self.client)
+                elif document_dict["path"]:
+                    doc_ref = DocumentReference(
+                        *document_dict["path"].split("/"), client=self.client
+                    )
+                else:
+                    continue
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:198]: In this case we were unable to get a valid Document Reference.

Either because, there's no collection set by Cx, nor valid doc_id (they could be sending empty strings or None), nor path in the document_dict.

We can't insert that document. We either fail all the insert (which could be thousands of documents) or skip it.
IIRC I mentioned I would skip it if this happened in the design doc.

I can raise an error if you prefer, I'm not certain what is better here. Some documents could be already committed in either case. And either the Cx tries to figure out what documents where inserted, then remove them and retry.

File: src/langchain_google_firestore/document_loader.py [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+            for elem in batch:
+                if document_ids:
+                    document_path = elem
+                else:
+                    document_dict = DocumentConverter.convertLangChainDocument(
+                        elem, self.client
+                    )
+                    document_path = document_dict["path"]
+                if not document_path:
+                    continue
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/document_loader.py:238]: Is the same as above. There is no DocumentReference and we can't delete the document. 
I can add an invalid path and should raise an error that `Document or collection path cannot be empty`.

I preferred to skip those documents and delete the valid documents. 
I can raise a condition if you think it's better.

File: src/langchain_google_firestore/utility/document_converter.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+if TYPE_CHECKING:
+    from google.cloud.firestore_v1.client import Client
+    from google.cloud.firestore_v1.base_document import DocumentSnapshot
+    from google.cloud.firestore_v1.document import DocumentReference
+
+
+class DocumentConverter:
+
+    @staticmethod
+    def convertFirestoreDocument(
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py]: In this case I'm not sure what's the better practice. This method does not require any member variable. It could be outside of the class as some package method.

I followed more of the static utility classes pattern but if you think it all of these should be without any class and just in this file I can change it.

File: src/langchain_google_firestore/utility/document_converter.py [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+        page_content_fields: List[str] = None,
+        metadata_fields: List[str] = None,
+    ) -> Document:
+        data_doc = document.to_dict()
+        metadata = {"reference": {"path": document.reference.path}}
+
+        if page_content_fields:
+            set_page_fields = set(page_content_fields)
+        else:
+            set_page_fields = set()
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py]: DONE

File: src/langchain_google_firestore/utility/document_converter.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py:39]: Sorry, ain't camel case starting lowercase letter ?

Pascal case would be  ConvertFirestoreDocument iiuc-- are you referring to this ?

I searched and looks like is already in camel case

camelCase
snake_case
PascalCase

File: src/langchain_google_firestore/utility/document_converter.py [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+        val_converted = val
+        if isinstance(val, dict):
+            l = len(val)
+            if (l == 1) and ("path" in val):
+                val_converted = DocumentReference(
+                    *val["path"].split("/"), client=client
+                )
+            elif (l == 2) and ("latitude" in val) and ("longitude" in val):
+                val_converted = GeoPoint(val["latitude"], val["longitude"])
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py:168]: Sorry could you elaborate on the `firestore_type` field?  

Do you mean that the GeoPoint --> Document  should add an extra key-value called "type":"firestore_type" ??

At the moment a conversion error will bubble up a user message if the parameters are incorrect

File: src/langchain_google_firestore/chat_message_history.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+import json
+
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import (
+    BaseMessage,
+    messages_from_dict,
+)
+
+USER_AGENT = "LangChain"
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py]: DONE

File: src/langchain_google_firestore/utility/document_converter.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_firestore/utility/document_converter.py]: As an alternative I could complicate the tests for this comparison. Either stop using the TestCase utility that simplifies most of the tests and add for loops that extracts and converts to dict and then compare. Or need to investigate how to add some comparison function and have a variant of TestCase method used.

File: src/langchain_google_firestore/utility/document_converter.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kurtisvg) [reply src/langchain_google_firestore/utility/document_converter.py:39]: Type on my part -- I meant "shouldn't use" 

which is another way of saying we should use snake_case

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: src/langchain_google_firestore/chat_message_history.py [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+                self.messages = MessageConverter.decode_messages(
+                    data_messages["messages"]
+                )
+
+    def add_message(self, message: BaseMessage) -> None:
+        self.messages.append(message)
+        self._upsert_messages()
+
+    def _upsert_messages(self) -> None:
+        self.doc_ref.set({"messages": MessageConverter.encode_messages(self.messages)})
```
Comment (wu-hui) [inline src/langchain_google_firestore/chat_message_history.py:74]: Just to confirm, `messages` is the only field of the document, and will stay this way right? Otherwise, other fields are getting overwritten.

**Review by wu-hui — APPROVED** [review]

File: src/langchain_google_firestore/chat_message_history.py (hunk 21 above)
Comment (JU-2094) [reply src/langchain_google_firestore/chat_message_history.py:74]: Yes, it's only `messages`

File: .github/workflows/lint.yml [hunk 22]
```diff
@@ -47,8 +47,9 @@ jobs:
 
       - name: Run linters
         run: |
-          black --check .
-          isort --check .
+          black --version
```
Comment (kurtisvg) [inline .github/workflows/lint.yml:50]: please revert these changes back

**Review by kurtisvg — APPROVED** [review]
